godot-kotlin/godot-library/src/nativeCore/kotlin/godot/core/type/PoolByteArray.kt

```diff
@@ -0,0 +1,166 @@
+        }
+    }
+
+
+    /**
+     * Appends a PoolByteArray at the end of this array.
+     */
+    fun appendArray(array: PoolByteArray) {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_append_array(it, array.handle.ptr))
+        }
+    }
+
+    /**
+     * Returns true if the array is empty.
+     */
+    fun empty() {
+        callNative {
+            checkNotNull(Godot.gdnative12.godot_pool_byte_array_empty(it))
+        }
+    }
+
+    /**
+     *  Retrieve the element at the given index.
+     */
+    operator fun get(idx: Int): UByte {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_get(it, idx))
+        }
+    }
+
+    /**
+     * Inserts a new element at a given position in the array.
+     * The position must be valid, or at the end of the array (idx == size()).
+     */
+    fun insert(idx: Int, data: UByte) {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_insert(it, idx, data))
+        }
+    }
+
+    /**
+     * Reverses the order of the elements in the array.
+     */
+    fun invert() {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_invert(it))
+        }
+    }
+
+    /**
+     * Appends a value to the array.
+     */
+    fun pushBack(data: UByte) {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_push_back(it, data))
+        }
+    }
+
+    /**
+     * Removes an element from the array by index.
+     */
+    fun remove(idx: Int) {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_remove(it, idx))
+        }
+    }
+
+    /**
+     * Sets the size of the array. If the array is grown, reserves elements at the end of the array.
+     * If the array is shrunk, truncates the array to the new size.
+     */
+    fun resize(size: Int) {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_resize(it, size))
+        }
+    }
+
+    /**
+     * Changes the Byte at the given index.
+     */
+    operator fun set(idx: Int, data: UByte) {
+        callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_set(it, idx, data))
+        }
+    }
+
+    /**
+     * Returns the size of the array.
+     */
+    fun size(): Int {
+        return callNative {
+            checkNotNull(Godot.gdnative.godot_pool_byte_array_size(it))
+        }
+    }
+
+    //UTILITIES
+    fun append(byte: Int) {
+        append(byte.toUByte())
+    }
+
+    operator fun set(idx: Int, byte: Int) {
+        this[idx] = byte.toUByte()
+    }
+
+    override fun toString(): String {
+        return "PoolByteArray(${size()})"
+    }
+
+    override fun iterator(): Iterator<UByte> {
+        return IndexedIterator(size(), this::get)
+    }
+
+    internal inline fun <T> callNative(block: MemScope.(CPointer<godot_pool_byte_array>) -> T): T {
```

Comment:
If I remember well, we have other cores that use Native calls. Maybe this method should go in internal sourceset. But I understand that it is not the point of this PR as we are discussing the template.

Reply:
Hard to move this up. I need to access the internal handle. It's not part of the CoreType interface (and because it is internal I can't add it. Furthermore not all the CoreTypes have this property anyway)

Reply:
I don't get it. It is not possible to make this method static in internal sourceset ?

Reply:
I can't use a static function for handling an instance property (_handle). It maybe be possible if the coretype itself is a parameter of the function but it doesn't work because _handle is not a property of the Coretype interface.
I tried to create a NativeCoreType abstract class that implement CoreType and add the _handle property but I still need to deal with the generic C structure. I tried something but couldn't around this issue even with reified type parameters.

`internal abstract class NativeCoreType<C : CStructVar> : CoreType {
    internal lateinit var _handle: CValue<C>
    
}

internal inline fun <T, reified C : CStructVar> callNative(
    nativeCore: NativeCoreType<C>,
    block: MemScope.(CPointer<C>) -> T
): T {
    return memScoped {
        val ptr = nativeCore._handle.ptr
        val ret: T = block(ptr)
        nativeCore._handle = ptr.pointed.readValue()
        ret
    }
}`

Reply:
I tried to get it work by defining nativeCall in abstract class int 923c6d8.
Maybe I'm missing something about the problem ?